File: src/lib/api/reading.ts
import { invoke } from "@tauri-apps/api/core";

export type FeedName = "top" | "best" | "new" | "ask" | "show" | "jobs";

export interface Story {
  id: number;
  title: string;
  url: string | null;
  domain: string | null;
  by: string;
  score: number;
  descendants: number;
  time: number;
  text: string | null;
  kind: string;
}

export interface Comment {
  id: number;
  author: string | null;
  html: string;
  text: string;
  created_at: string;
  depth: number;
  children: Comment[];
  subtree_size: number;
}

export interface Thread {
  id: number;
  title: string;
  url: string | null;
  domain: string | null;
  author: string | null;
  points: number | null;
  created_at: string;
  text: string | null;
  comments: Comment[];
  comment_count: number;
}

export interface ThreadView {
  thread: Thread;
  newComments: number | null;
  lastVisit: number | null;
}

export interface Article {
  url: string;
  title: string;
  byline: string | null;
  site_name: string | null;
  excerpt: string | null;
  published_time: string | null;
  markdown: string;
  word_count: number;
  note: string | null;
}

export interface Coverage {
  included: number;
  total: number;
  chars: number;
}

export function loadFeed(feed: FeedName, offset: number, limit: number) {
  return invoke<Story[]>("feed", { feed, offset, limit });
}

export function searchStories(query: string, byDate: boolean) {
  return invoke<Story[]>("search_stories", { query, byDate });
}

export function loadThread(id: number, refresh = false) {
  return invoke<ThreadView>("load_thread", { id, refresh });
}

export function loadArticle(url: string, refresh = false) {
  return invoke<Article>("load_article", { url, refresh });
}

export function readIds() {
  return invoke<number[]>("read_ids");
}

export function coverage(storyId: number) {
  return invoke<Coverage>("coverage", { storyId });
}

export function resolveItem(id: number) {
  return invoke<{ storyId: number; commentId: number | null }>("resolve_item", { id });
}

File: src-tauri/src/commands/reading.rs
use crate::article::{self, Article};
use crate::hn::{self, ItemRef, Story, Thread};
use crate::prompts;
use crate::store::Store;
use crate::{fail, Fallible, ARTICLE_TTL, THREAD_TTL};
use serde::Serialize;
use tauri::State;

#[tauri::command]
pub(crate) async fn feed(feed: String, offset: usize, limit: usize) -> Fallible<Vec<Story>> {
    hn::feed_page(&feed, offset, limit).await.map_err(fail)
}

#[tauri::command]
pub(crate) async fn search_stories(query: String, by_date: bool) -> Fallible<Vec<Story>> {
    hn::search(&query, by_date).await.map_err(fail)
}

pub(crate) async fn resolve_thread(store: &Store, id: u64, refresh: bool) -> anyhow::Result<Thread> {
    let key = id.to_string();
    if !refresh {
        if let Ok(Some(raw)) = store.cache_get("thread", &key, THREAD_TTL) {
            if let Ok(thread) = serde_json::from_str::<Thread>(&raw) {
                return Ok(thread);
            }
        }
    }
    let thread = hn::thread(id).await?;
    if let Ok(raw) = serde_json::to_string(&thread) {
        let _ = store.cache_put("thread", &key, &raw);
    }

    // The whole discussion, so a half-remembered comment is findable later.
    let body = hn::flatten(&thread.comments)
        .iter()
        .map(|comment| {
            format!(
                "{}: {}",
                comment.author.as_deref().unwrap_or("unknown"),
                comment.text
            )
        })
        .collect::<Vec<_>>()
        .join("\n\n");
    let _ = store.library_put(id, &thread.title, "thread", &body);

    Ok(thread)
}

pub(crate) async fn resolve_article(store: &Store, url: &str, refresh: bool) -> anyhow::Result<Article> {
    if !refresh {
        if let Ok(Some(raw)) = store.cache_get("article", url, ARTICLE_TTL) {
            if let Ok(article) = serde_json::from_str::<Article>(&raw) {
                return Ok(article);
            }
        }
    }
    let article = article::extract(url).await?;
    if let Ok(raw) = serde_json::to_string(&article) {
        let _ = store.cache_put("article", url, &raw);
    }
    Ok(article)
}

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub(crate) struct ThreadView {
    thread: Thread,
    /// Comments added since this thread was last opened. `None` on a first
    /// visit, which the UI shows differently from "nothing new".
    new_comments: Option<u32>,
    /// Unix seconds of the previous visit, so individual comments posted since
    /// then can be marked in the tree.
    last_visit: Option<i64>,
}

#[tauri::command]
pub(crate) async fn load_thread(store: State<'_, Store>, id: u64, refresh: bool) -> Fallible<ThreadView> {
    let thread = resolve_thread(&store, id, refresh)
        .await
        .map_err(|err| format!("Could not load this thread: {err}"))?;
    let previous = store.visit(id, thread.comment_count).unwrap_or(None);
    let new_comments = previous.map(|(before, _)| thread.comment_count.saturating_sub(before));
    let last_visit = previous.map(|(_, read_at)| read_at);
    Ok(ThreadView {
        thread,
        new_comments,
        last_visit,
    })
}

#[tauri::command]
pub(crate) async fn load_article(store: State<'_, Store>, url: String, refresh: bool) -> Fallible<Article> {
    resolve_article(&store, &url, refresh).await.map_err(fail)
}

#[tauri::command]
pub(crate) fn read_ids(store: State<'_, Store>) -> Fallible<Vec<u64>> {
    store.read_ids().map_err(fail)
}

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub(crate) struct Coverage {
    included: usize,
    total: usize,
    chars: usize,
}

#[tauri::command]
pub(crate) async fn coverage(store: State<'_, Store>, story_id: u64) -> Fallible<Coverage> {
    let thread = resolve_thread(&store, story_id, false).await.map_err(fail)?;
    let article = match thread.url.as_deref() {
        Some(url) => resolve_article(&store, url, false).await.ok(),
        None => None,
    };
    let stats = prompts::pack_stats(&thread, article.as_ref());
    Ok(Coverage {
        included: stats.included,
        total: stats.total,
        chars: stats.chars,
    })
}

/// Opens whatever a pasted Hacker News link points at.
#[tauri::command]
pub(crate) async fn resolve_item(id: u64) -> Fallible<ItemRef> {
    hn::resolve_item(id)
        .await
        .map_err(|err| format!("Could not open that link: {err}"))
}
